### PR TITLE
Python: Add support for magic methods

### PR DESCRIPTION
**Pull request by yoff**
This PR labels certain control flow nodes as corresponding to method calls to magic methods. This is then picked up by the data flow library to resolve flow between subexpressions and parameters.

Currently only binary operators are supported and only the standard form (not the r-variant).

### CODE REVIEW

**Comment by yoff** [comment]
Ah, I forgot to annotate the test file with the newly obtained results..

**Comment by yoff** [comment]
There were no annotations present, so I did it in two steps to make the diff clear. The first commit marks all calls as not found, the second one unmarks the ones found by the new magic feature.

File: python/ql/src/experimental/dataflow/internal/DataFlowPrivate.qll [hunk 1]
```diff
@@ -157,17 +158,67 @@ class DataFlowClassValue extends DataFlowCallable, TClassValue {
...
-  DataFlowCall() { this = callable.getACall() }
+abstract class DataFlowCall extends TDataFlowCall {
+  /** Gets a textual representation of this element. */
+  abstract string toString();
 
   /** Get the callable to which this call goes. */
-  DataFlowCallable getCallable() { result = callable }
+  abstract DataFlowCallable getCallable();
+
+  /** Get the specified arguemnt to this call. */
```
Comment (intrigus-lgtm) [inline python/ql/src/experimental/dataflow/internal/DataFlowPrivate.qll:172]:
```suggestion
  /** Get the specified argument to this call. */
```

File: python/ql/src/experimental/dataflow/internal/DataFlowPrivate.qll [hunk 2]
```diff
@@ -157,17 +158,67 @@ class DataFlowClassValue extends DataFlowCallable, TClassValue {
...
+  /** Get the control flow node representing this call. */
+  abstract ControlFlowNode getNode();
 
   /** Gets the enclosing callable of this call. */
-  DataFlowCallable getEnclosingCallable() { result.getScope() = this.getNode().getScope() }
+  abstract DataFlowCallable getEnclosingCallable();
+}
+
+
+/** Represents a call to a callable */
```
Comment (intrigus-lgtm) [inline python/ql/src/experimental/dataflow/internal/DataFlowPrivate.qll:183]:
```suggestion
/** Represents a call to a callable. */
```

File: python/ql/src/experimental/dataflow/internal/DataFlowPrivate.qll [hunk 3]
```diff
@@ -157,17 +158,67 @@ class DataFlowClassValue extends DataFlowCallable, TClassValue {
...
+  }
+
+  override ControlFlowNode getNode() { result = call }
+
+  override DataFlowCallable getCallable() { result = callable }
+
+  override DataFlowCallable getEnclosingCallable() { result.getScope() = call.getNode().getScope() }
+}
+
+class MagicCall extends DataFlowCall, TMagicCall {
```
Comment (intrigus-lgtm) [inline python/ql/src/experimental/dataflow/internal/DataFlowPrivate.qll:206]: Missing QlDoc?

File: python/ql/src/semmle/python/Magic.qll [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+import python
+
+/**
+ * Machinery for detecting magic method calls.
+ * Extend `MagicMethod::Potential` to capture more cases.
+ */
+module MagicMethod {
+  /** A control flow node which might correpsond to a magic method call. */
+  abstract class Potential extends ControlFlowNode {
+    /** Gets the name of the method that would be called */
```
Comment (intrigus-lgtm) [inline python/ql/src/semmle/python/Magic.qll:19]:
```suggestion
    /** Gets the name of the method that would be called. */
```

File: python/ql/src/semmle/python/Magic.qll [hunk 5]
```diff
@@ -0,0 +1,38 @@
+import python
+
+module MagicMethod {
```
[the remark this answers is not in the document]
Comment (yoff) [reply python/ql/src/semmle/python/Magic.qll]: It has been remedied :-)

File: python/ql/src/semmle/python/Magic.qll [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+  }
+
+  /** A control flow node corresponding to a magic method call. */
+  class Actual extends ControlFlowNode {
+    Value resolvedMagicMethod;
+
+    Actual() {
+      exists(Potential pot |
+        this.(Potential) = pot and
+        pot.getSelf().pointsTo().getClass().lookup(pot.getMagicMethodName()) = resolvedMagicMethod
```
[the remark this answers is not in the document]
Comment (yoff) [reply python/ql/src/semmle/python/Magic.qll]: I have adopted the above structure.

File: python/ql/src/experimental/dataflow/internal/DataFlowPrivate.qll (hunk 3 above)
Comment (yoff) [reply python/ql/src/experimental/dataflow/internal/DataFlowPrivate.qll]: Yes, added.

File: python/ql/src/semmle/python/SpecialMethods.qll [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+ *   - A subset of `ControlFlowNode`s are labelled as potentially corresponding to
+ *     a special method call (by being an instance of `SpecialMethod::Potential`).
+ *   - A subset of the potential special method calls are labelled as being actual
+ *     special method calls (`SpecialMethodCallNode`) if the appropriate method is defined.
+ * Extend `SpecialMethod::Potential` to capture more cases.
+ */
+
+import python
+
+/** A control flow node which might correpsond to a special method call. */
```
Comment (tausbn) [inline python/ql/src/semmle/python/SpecialMethods.qll:13]:
```suggestion
/** A control flow node which might correspond to a special method call. */
```

File: python/ql/src/semmle/python/SpecialMethods.qll [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+class PotentialSpecialMethodCallNode extends ControlFlowNode {
+  PotentialSpecialMethodCallNode() { this instanceof SpecialMethod::Potential }
+}
+
+/**
+ * Machinery for detecting special method calls.
+ * Extend `SpecialMethod::Potential` to capture more cases.
+ */
+module SpecialMethod {
+  /** A control flow node which might correpsond to a special method call. */
```
Comment (tausbn) [inline python/ql/src/semmle/python/SpecialMethods.qll:23]:
```suggestion
  /** A control flow node which might correspond to a special method call. */
```

File: python/ql/src/semmle/python/SpecialMethods.qll [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+  abstract class Potential extends ControlFlowNode {
+    /** Gets the name of the method that would be called. */
+    abstract string getSpecialMethodName();
+
+    /** Gets the controlflow node that would be passed as the specified argument. */
```
Comment (tausbn) [inline python/ql/src/semmle/python/SpecialMethods.qll:28]:
```suggestion
    /** Gets the control flow node that would be passed as the specified argument. */
```

**Review by tausbn — CHANGES_REQUESTED** [review]
Found a few typos. Otherwise looks good to me. :+1:

**Review by tausbn — APPROVED** [review]

**Comment by tausbn** [comment]
I think this looks good to go, once the merge conflicts have been ironed out.

**Comment by yoff** [comment]
"dismissed"?! I Specifically merged main to iron out those conflicts...

**Comment by RasmusWL** [comment]
Errors on jenkins `ERROR: The cast type DataFlowCall is incompatible with the expression type ControlFlowNode. (argumentRouting1.ql:12,23-45)`

**Review by tausbn — APPROVED** [review]

**Pull request merged**